agent-jvm/src/main/java/org/jolokia/mcp/jvmagent/InVmRequestHandler.java

```diff
@@ -0,0 +1,386 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jolokia.mcp.jvmagent;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import javax.management.AttributeNotFoundException;
+import javax.management.InstanceNotFoundException;
+import javax.management.JMException;
+import javax.management.MBeanException;
+import javax.management.ReflectionException;
+import javax.management.RuntimeMBeanException;
+
+import org.jolokia.json.JSONArray;
+import org.jolokia.json.JSONObject;
+import org.jolokia.json.JSONStructure;
+import org.jolokia.json.parser.JSONParser;
+import org.jolokia.json.parser.ParseException;
+import org.jolokia.server.core.backend.BackendManager;
+import org.jolokia.server.core.config.ConfigKey;
+import org.jolokia.server.core.request.BadRequestException;
+import org.jolokia.server.core.request.EmptyResponseException;
+import org.jolokia.server.core.request.JolokiaExecRequest;
+import org.jolokia.server.core.request.JolokiaListRequest;
+import org.jolokia.server.core.request.JolokiaReadRequest;
+import org.jolokia.server.core.request.JolokiaRequest;
+import org.jolokia.server.core.request.JolokiaRequestFactory;
+import org.jolokia.server.core.request.JolokiaWriteRequest;
+import org.jolokia.server.core.request.ProcessingParameters;
+import org.jolokia.server.core.service.api.JolokiaContext;
+import org.jolokia.server.core.util.MimeTypeUtil;
+import org.jolokia.server.core.util.RequestType;
+
+public class InVmRequestHandler {
```

Comment:
Looks like this class has a lot in common with `org.jolokia.server.core.http.HttpRequestHandler` which also doesn't implement any special interface (nor doesn't extend any special abstract class).
At some point we may extract common functionality like parameter validation.

Reply:
Yes, that's what I thought as well! If we agree to go for this path, next we can consider providing an invm generic request handler so that Jolokia itself and users as well can invoke the Jolokia functionalities invm. I'll file an GH issue to the jolokia repo.

Reply:
https://github.com/jolokia/jolokia/issues/856